### DISCUSSION TITLE
How to debug mapperly in Visual Studio 2022?

### DISCUSSION THREAD

**Discussion by ycherkes** [riok/mapperly#50] [Q&A] [answered]
Hi, I've cloned mapperly, created one more Console project, added two project references:

    <ItemGroup>
        <ProjectReference Include="..\Riok.Mapperly.Abstractions\Riok.Mapperly.Abstractions.csproj" />
        <ProjectReference Include="..\Riok.Mapperly\Riok.Mapperly.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" />
    </ItemGroup>

After these manipulations, mapperly generates all the mappers, but I can't start debugging mapperly.
I've tried: https://github.com/JoanComasFdz/dotnet-how-to-debug-source-generator-vs2022 and https://github.com/dotnet/roslyn-sdk/issues/850 with no success.

Could you help me?

**Comment by latonz** [comment, marked as answer]
Note the compilation and therefore the execution of the source generator happens in a different process, before the command line application runs.

Mapperly uses a compile constant `DEBUG_SOURCE_GENERATOR` to determine whether to attach a debugger. If it is set, it tries to attach a debugger (it uses `Debugger.Launch()` on windows, on macos it tries to launch jetbrains rider) and waits for the connection of the debugger (up to 30seconds). If the automatic debugger attachment fails (eg. if rider can't be found on macos, you use linux, ...) you can use these 30 seconds to attach your debugger manually (for vs 2022 see [attach to running processes](https://docs.microsoft.com/en-us/visualstudio/debugger/attach-to-running-processes-with-the-visual-studio-debugger?view=vs-2022)). You can use the `DefineConstants` dontet build or csproj parameter to define the `DEBUG_SOURCE_GENERATOR` constant or temporary comment the conditional attribute on `DebuggerUtil.AttachDebugger`.

**Reply by ycherkes** [reply]
Thanks a lot.
First - as I have Console app targeted to .net6, I saw errors: "An instance of analyzer cannot be created".
It's fixed by adding **SetTargetFramework** attribute to the target console project file: 

<ProjectReference Include="..\Riok.Mapperly\Riok.Mapperly.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" **SetTargetFramework="TargetFramework=netstandard2.0"**/>

The solution was found on [stackoverflow](https://stackoverflow.com/a/66702383/7901167)

Second - I've defined a DEBUG_SOURCE_GENERATOR debug constant in a Riok.Mapperly project file as you said.
As result, I'm able to debug the mapperly.

**Comment by MirzaMerdovic** [comment]
Hi,

Thanks for the info here! but it might be good to add this as a part of readme or the official documentation.

**Reply by latonz** [reply]
I created https://github.com/riok/mapperly/issues/231 to track this.
